### PR TITLE
colmap: fix for cuda11.6/gcc11

### PR DESCRIPTION
**Pull request by SomeoneSerge**
###### Description of changes

Colmap has been failing to build with cuda after `gcc10->gcc11` bump:

- [With cuda](https://hercules-ci.com/accounts/github/SomeoneSerge/derivations/%2Fnix%2Fstore%2Fvfpg5h9rmdikrzp5ny64dvxqdw0vcgxl-colmap-3.7.drv/log?via-job=d09bd716-8357-4ffc-b46e-6069c7042591)
- And, I guess, [without it](https://hydra.nixos.org/build/174170349)

This PR introduces a temporary fix, taken from https://github.com/colmap/colmap/issues/1418#issuecomment-1049305256

I'm still waiting for the builds to finish

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

CC @NixOS/cuda-maintainers 
CC @lebastr

### CODE REVIEW

File: pkgs/applications/science/misc/colmap/default.nix [hunk 1]
```diff
@@ -1,33 +1,49 @@
...
   pname = "colmap";
   src = fetchFromGitHub {
-     owner = "colmap";
-     repo = "colmap";
-     rev = version;
-     sha256 = "sha256-uVAw6qwhpgIpHkXgxttKupU9zU+vD0Za0maw2Iv4x+I=";
+    owner = "colmap";
+    repo = "colmap";
+    rev = version;
+    sha256 = "sha256-uVAw6qwhpgIpHkXgxttKupU9zU+vD0Za0maw2Iv4x+I=";
```
Comment (samuela) [inline pkgs/applications/science/misc/colmap/default.nix:18]: nit: use hash instead of sha256

**Review by samuela — APPROVED** [review]
nit: formatting diffs are mixed in with semantic diffs, making this a little tricky to read

overall LGTM!

**Comment by SomeoneSerge** [comment]
@samuela ouch, I actually tried to preserve formatting; maybe I've accidentally pressed `\F`

**Comment by SomeoneSerge** [comment]
Ouch, actually the build still fails. Will look later

**Comment by samuela** [comment]
Result of `nixpkgs-review pr 169623` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages failed to build:</summary>
  <ul>
    <li>colmap</li>
    <li>colmapWithCuda</li>
  </ul>
</details>

**Comment by samuela** [comment]
I'm seeing some build failures:
```
error: builder for '/nix/store/cz0ng75yp2xvcdm9pfjpq3z5clr6panz-colmap-3.7.drv' failed with exit code 2;
       last 10 log lines:
       > /nix/store/m2vh2ny7bqpwij1gpmvl5gxj7y4dgr4f-binutils-2.38/bin/ld: (.text+0x5d8a): undefined reference to `std::__throw_bad_array_new_length()'
       > /nix/store/m2vh2ny7bqpwij1gpmvl5gxj7y4dgr4f-binutils-2.38/bin/ld: /nix/store/22wgdfnkh18gva3aw2ymfzyb44z1pfrc-boost-1.77.0/lib/libboost_program_options-mt-x64.a(cmdline.o): in function `boost::program_options::detail::cmdline::run()':
       > (.text+0x6c76): undefined reference to `std::__throw_bad_array_new_length()'
       > /nix/store/m2vh2ny7bqpwij1gpmvl5gxj7y4dgr4f-binutils-2.38/bin/ld: (.text+0x6c84): undefined reference to `std::__throw_bad_array_new_length()'
       > /nix/store/m2vh2ny7bqpwij1gpmvl5gxj7y4dgr4f-binutils-2.38/bin/ld: /nix/store/22wgdfnkh18gva3aw2ymfzyb44z1pfrc-boost-1.77.0/lib/libboost_program_options-mt-x64.a(cmdline.o):(.text+0x7ab2): more undefined references to `std::__throw_bad_array_new_length()' follow
       > /nix/store/m2vh2ny7bqpwij1gpmvl5gxj7y4dgr4f-binutils-2.38/bin/ld: /nix/store/33fy01bvz2s4fq7nb536mbvcy1ifdknz-glog-0.6.0/lib/libglog.so: undefined reference to `std::__throw_bad_array_new_length()@GLIBCXX_3.4.29'
       > collect2: error: ld returned 1 exit status
       > make[2]: *** [src/exe/CMakeFiles/colmap_exe.dir/build.make:251: src/exe/colmap] Error 1
       > make[1]: *** [CMakeFiles/Makefile2:775: src/exe/CMakeFiles/colmap_exe.dir/all] Error 2
       > make: *** [Makefile:136: all] Error 2
       For full logs, run 'nix log /nix/store/cz0ng75yp2xvcdm9pfjpq3z5clr6panz-colmap-3.7.drv'.
error: builder for '/nix/store/w6gr8cjy42n5y544nxpfx9vcasm2k6aw-colmap-3.7.drv' failed with exit code 2;
       last 10 log lines:
       > (.text+0x5d80): undefined reference to `std::__throw_bad_array_new_length()'
       > /nix/store/m2vh2ny7bqpwij1gpmvl5gxj7y4dgr4f-binutils-2.38/bin/ld: (.text+0x5d8a): undefined reference to `std::__throw_bad_array_new_length()'
       > /nix/store/m2vh2ny7bqpwij1gpmvl5gxj7y4dgr4f-binutils-2.38/bin/ld: /nix/store/22wgdfnkh18gva3aw2ymfzyb44z1pfrc-boost-1.77.0/lib/libboost_program_options-mt-x64.a(cmdline.o): in function `boost::program_options::detail::cmdline::run()':
       > (.text+0x6c76): undefined reference to `std::__throw_bad_array_new_length()'
       > /nix/store/m2vh2ny7bqpwij1gpmvl5gxj7y4dgr4f-binutils-2.38/bin/ld: /nix/store/22wgdfnkh18gva3aw2ymfzyb44z1pfrc-boost-1.77.0/lib/libboost_program_options-mt-x64.a(cmdline.o):(.text+0x6c84): more undefined references to `std::__throw_bad_array_new_length()' follow
       > /nix/store/m2vh2ny7bqpwij1gpmvl5gxj7y4dgr4f-binutils-2.38/bin/ld: /nix/store/33fy01bvz2s4fq7nb536mbvcy1ifdknz-glog-0.6.0/lib/libglog.so: undefined reference to `std::__throw_bad_array_new_length()@GLIBCXX_3.4.29'
       > collect2: error: ld returned 1 exit status
       > make[2]: *** [src/exe/CMakeFiles/colmap_exe.dir/build.make:256: src/exe/colmap] Error 1
       > make[1]: *** [CMakeFiles/Makefile2:803: src/exe/CMakeFiles/colmap_exe.dir/all] Error 2
       > make: *** [Makefile:136: all] Error 2
       For full logs, run 'nix log /nix/store/w6gr8cjy42n5y544nxpfx9vcasm2k6aw-colmap-3.7.drv'.
error: 2 dependencies of derivation '/nix/store/5fiaywf2vncf60gyd4vlpazglmlxiqvb-review-shell.drv' failed to build
```

**Comment by SomeoneSerge** [comment]
Yes, these are the same errors I was trying to fix. I'm now retrying with:

```
"-DCMAKE_CUDA_STANDARD=14"
```

**Comment by SomeoneSerge** [comment]
Also, it seems it's still using gcc10 - probably because of the pinned minor version of qt

**Comment by samuela** [comment]
Result of `nixpkgs-review pr 169623` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages failed to build:</summary>
  <ul>
    <li>colmap</li>
    <li>colmapWithCuda</li>
  </ul>
</details>

**Comment by samuela** [comment]
new failure log:
```
error: builder for '/nix/store/jjv1jdvs0w8c9j5hz5l071apmlnhym3k-colmap-3.7.drv' failed with exit code 2;
       last 10 log lines:
       > /nix/store/m2vh2ny7bqpwij1gpmvl5gxj7y4dgr4f-binutils-2.38/bin/ld: (.text+0x5d8a): undefined reference to `std::__throw_bad_array_new_length()'
       > /nix/store/m2vh2ny7bqpwij1gpmvl5gxj7y4dgr4f-binutils-2.38/bin/ld: /nix/store/22wgdfnkh18gva3aw2ymfzyb44z1pfrc-boost-1.77.0/lib/libboost_program_options-mt-x64.a(cmdline.o): in function `boost::program_options::detail::cmdline::run()':
       > (.text+0x6c76): undefined reference to `std::__throw_bad_array_new_length()'
       > /nix/store/m2vh2ny7bqpwij1gpmvl5gxj7y4dgr4f-binutils-2.38/bin/ld: (.text+0x6c84): undefined reference to `std::__throw_bad_array_new_length()'
       > /nix/store/m2vh2ny7bqpwij1gpmvl5gxj7y4dgr4f-binutils-2.38/bin/ld: /nix/store/22wgdfnkh18gva3aw2ymfzyb44z1pfrc-boost-1.77.0/lib/libboost_program_options-mt-x64.a(cmdline.o):(.text+0x7ab2): more undefined references to `std::__throw_bad_array_new_length()' follow
       > /nix/store/m2vh2ny7bqpwij1gpmvl5gxj7y4dgr4f-binutils-2.38/bin/ld: /nix/store/33fy01bvz2s4fq7nb536mbvcy1ifdknz-glog-0.6.0/lib/libglog.so: undefined reference to `std::__throw_bad_array_new_length()@GLIBCXX_3.4.29'
       > collect2: error: ld returned 1 exit status
       > make[2]: *** [src/exe/CMakeFiles/colmap_exe.dir/build.make:251: src/exe/colmap] Error 1
       > make[1]: *** [CMakeFiles/Makefile2:775: src/exe/CMakeFiles/colmap_exe.dir/all] Error 2
       > make: *** [Makefile:136: all] Error 2
       For full logs, run 'nix log /nix/store/jjv1jdvs0w8c9j5hz5l071apmlnhym3k-colmap-3.7.drv'.
[1/0/2 built (1 failed), 0.0 MiB DL] building colmap-3.7 (buildPhase): (1215): herethread 'tokio-runtime-worker' panicked at 'borked calling describe_instance_status: HttpDispatch(HttpDispatchError { message: "Error during dispatch: connection closed before message completed" })', src/main.rs:55:8
                                                                                                  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
error: builder for '/nix/store/1pdrpxz7f1n0gxrcnnw4sqvynmqrhggl-colmap-3.7.drv' failed with exit code 2;
       last 10 log lines:
       >   537 |       static constexpr intmax_t den = type::den;
       >       |                                  ^~~
       > CMake Error at colmap_cuda_generated_patch_match_cuda.cu.o.Release.cmake:280 (message):
       >   Error generating file
       >   /build/source/build/src/CMakeFiles/colmap_cuda.dir/mvs/./colmap_cuda_generated_patch_match_cuda.cu.o
       >
       >
       > make[2]: *** [src/CMakeFiles/colmap_cuda.dir/build.make:91: src/CMakeFiles/colmap_cuda.dir/mvs/colmap_cuda_generated_patch_match_cuda.cu.o] Error 1
       > make[1]: *** [CMakeFiles/Makefile2:768: src/CMakeFiles/colmap_cuda.dir/all] Error 2
       > make: *** [Makefile:136: all] Error 2
       For full logs, run 'nix log /nix/store/1pdrpxz7f1n0gxrcnnw4sqvynmqrhggl-colmap-3.7.drv'.
error: 2 dependencies of derivation '/nix/store/xl46g31f9ykhvsbiv6y3zr60xpim28gk-review-shell.drv' failed to build
```

**Comment by SomeoneSerge** [comment]
It's the same `std::__throw_bad_array_new_length()@GLIBCXX_3.4.29` thing, I'm sure it's coming from mixing gcc10 and gcc11 (I'll be trying to unpin qt minor version)

**Comment by SomeoneSerge** [comment]
Nope. Maybe I need to set CXX standard too 🤔

**Comment by gebner** [comment]
See also #170061, which upgrades colmap to Qt 5.15 and fixes the non-CUDA build.

**Comment by SomeoneSerge** [comment]
Ok, let's see [if the Qt PR fixes the colmap issue](https://hercules-ci.com/accounts/github/SomeoneSerge/derivations/%2Fnix%2Fstore%2Fzcg7dzn3wwh64d73jbwg0qskwf24w1yb-colmap-3.7.drv/log?via-job=c7e2f5ab-c383-4073-ad5f-b0abcac30fa4)

**Comment by SomeoneSerge** [comment]
1. https://github.com/NixOS/nixpkgs/pull/170061 has indeed fixed the non-cuda build
2. For CUDA build to work one _does_ need to force the nvcc arguments. Variables like `CMAKE_CUDA_STANDARD` don't seem to have any effect

I've rebased and autosquashed the branch, think it's ready

**Comment by samuela** [comment]
Result of `nixpkgs-review pr 169623` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>colmap</li>
    <li>colmapWithCuda</li>
  </ul>
</details>

File: pkgs/applications/science/misc/colmap/default.nix [hunk 2]
```diff
@@ -1,10 +1,14 @@
...
   freeimage, glog, libGLU, glew, qtbase,
-  cudaSupport ? false, cudatoolkit ? null }:
+  cudaSupport ? false, cudaPackages }:
 
-assert !cudaSupport || cudatoolkit != null;
+assert cudaSupport -> cudaPackages != { };
 
-let boost_static = boost17x.override { enableStatic = true; };
+let
+  boost_static = boost17x.override { enableStatic = true; };
```
Comment (samuela) [inline pkgs/applications/science/misc/colmap/default.nix:8]: Why do we need `enableStatic`? Why wasn't this necessary before?

Comment (samuela) [reply pkgs/applications/science/misc/colmap/default.nix:8]: oh oops just saw that this was necessary before... wait why is it necessary? does clomp support building against shared boost libs?

Comment (SomeoneSerge) [reply pkgs/applications/science/misc/colmap/default.nix:8]: I had the same questions, but decided not to touch it for now. I don't know why it uses static. Although static is by default preferable to dynamic. I also don't know why it pins a narrow range for boost versions

Comment (samuela) [reply pkgs/applications/science/misc/colmap/default.nix:8]: > Although static is by default preferable to dynamic.

Why is this? In Nix, shared libraries seem much more space efficient while still locking down the exact library versions. What's the benefit to static libraries?

Comment (SomeoneSerge) [reply pkgs/applications/science/misc/colmap/default.nix:8]: Right, I guess "by default" is plain incorrect wording. There is "a perspective", from which static builds are still "preferable": you get _fewer_ derivations in the runtime closure, might even get binaries that you can patchelf and ship to a non-nixos system, might even get a smaller runtime closure in some corner cases, when the expression shipping shared libraries also propagates irrelevant. For example, our `symlinkJoin`-ed cuda is a runtime dependency that ships shared libraries, but also nvcc and static archives: nix does not have means to know it won't need them at runtime. If all of cuda was linked statically, that wouldn't happen

Comment (lebastr) [reply pkgs/applications/science/misc/colmap/default.nix:8]: I have already forgotten the reason for this decision. It seems the comap build system required static boost

Comment (samuela) [reply pkgs/applications/science/misc/colmap/default.nix:8]: Yeah, I'd be curious to track that down. But that's out of scope for this PR

**Review by samuela — APPROVED** [review]
happy to merge whenever you feel ready @SomeoneSerge

**Comment by SomeoneSerge** [comment]
@samuela, good to go!

**Comment by SomeoneSerge** [comment]
Wait! I think I forgot `/run/opengl-driver/lib`

**Comment by samuela** [comment]
Result of `nixpkgs-review pr 169623` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>colmap</li>
    <li>colmapWithCuda</li>
  </ul>
</details>

**Comment by samuela** [comment]
LGTM ready to merge?

**Comment by SomeoneSerge** [comment]
Yup

**Comment by SomeoneSerge** [comment]
Btw, thanks @lebastr for bringing my attention to the missing rpath to the cuda driver!

**Pull request merged**